internal/storage/v2/api/tracestore/tracestoremetrics/reader_metrics.go

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2025 The Jaeger Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package tracestoremetrics
+
+import (
+	"context"
+	"iter"
+	"time"
+
+	"go.opentelemetry.io/collector/pdata/ptrace"
+
+	"github.com/jaegertracing/jaeger/internal/metrics"
+	"github.com/jaegertracing/jaeger/internal/storage/v2/api/tracestore"
+)
+
+var _ tracestore.Reader = (*ReadMetricsDecorator)(nil)
+
+// ReadMetricsDecorator wraps a tracestore.Reader and collects metrics around each read operation.
+type ReadMetricsDecorator struct {
+	traceReader          tracestore.Reader
+	findTracesMetrics    *queryMetrics
+	findTraceIDsMetrics  *queryMetrics
+	getTraceMetrics      *queryMetrics
+	getServicesMetrics   *queryMetrics
+	getOperationsMetrics *queryMetrics
+}
+
+type queryMetrics struct {
+	Errors     metrics.Counter `metric:"requests" tags:"result=err"`
+	Successes  metrics.Counter `metric:"requests" tags:"result=ok"`
+	Responses  metrics.Timer   `metric:"responses"` // used as a histogram, not necessary for GetTrace
+	ErrLatency metrics.Timer   `metric:"latency" tags:"result=err"`
+	OKLatency  metrics.Timer   `metric:"latency" tags:"result=ok"`
+}
+
+func (q *queryMetrics) emit(err error, latency time.Duration, responses int) {
+	if err != nil {
+		q.Errors.Inc(1)
+		q.ErrLatency.Record(latency)
+	} else {
+		q.Successes.Inc(1)
+		q.OKLatency.Record(latency)
+		q.Responses.Record(time.Duration(responses))
+	}
+}
+
+// NewReaderDecorator returns a new ReadMetricsDecorator.
+func NewReaderDecorator(traceReader tracestore.Reader, metricsFactory metrics.Factory) *ReadMetricsDecorator {
+	return &ReadMetricsDecorator{
+		traceReader:          traceReader,
+		findTracesMetrics:    buildQueryMetrics("find_traces", metricsFactory),
+		findTraceIDsMetrics:  buildQueryMetrics("find_trace_ids", metricsFactory),
+		getTraceMetrics:      buildQueryMetrics("get_trace", metricsFactory),
+		getServicesMetrics:   buildQueryMetrics("get_services", metricsFactory),
+		getOperationsMetrics: buildQueryMetrics("get_operations", metricsFactory),
+	}
+}
+
+func buildQueryMetrics(operation string, metricsFactory metrics.Factory) *queryMetrics {
+	qMetrics := &queryMetrics{}
+	scoped := metricsFactory.Namespace(metrics.NSOptions{Name: "", Tags: map[string]string{"operation": operation}})
+	metrics.Init(qMetrics, scoped, nil)
+	return qMetrics
+}
+
+// FindTraces implements tracestore.Reader#FindTraces
+func (m *ReadMetricsDecorator) FindTraces(ctx context.Context, query tracestore.TraceQueryParams) iter.Seq2[[]ptrace.Traces, error] {
+	return func(yield func([]ptrace.Traces, error) bool) {
+		start := time.Now()
+		findTracesIter := m.traceReader.FindTraces(ctx, query)
+		for traces, err := range findTracesIter {
+			m.findTracesMetrics.emit(err, time.Since(start), len(traces))
```

Comment:
this is incorrect, it will emit multiple times per one call to FindTraces.

you only need to emit once when returning from the function. Since there are more than one exit points, you can use defer to run emit exactly once. You will probably want to capture error in a var that will live for the duration of the function and be used in deferred function.